### PR TITLE
[SPARK-14267] [SQL] [PYSPARK] execute multiple Python UDFs within single batch

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This PR support multiple Python UDFs within single batch, also improve the performance.

``` python
>>> from pyspark.sql.types import IntegerType
>>> sqlContext.registerFunction("double", lambda x: x * 2, IntegerType())
>>> sqlContext.registerFunction("add", lambda x, y: x + y, IntegerType())
>>> sqlContext.sql("SELECT double(add(1, 2)), add(double(2), 1)").explain(True)
== Parsed Logical Plan ==
'Project [unresolvedalias('double('add(1, 2)), None),unresolvedalias('add('double(2), 1), None)]
+- OneRowRelation$

== Analyzed Logical Plan ==
double(add(1, 2)): int, add(double(2), 1): int
Project [double(add(1, 2))#14,add(double(2), 1)#15]
+- Project [double(add(1, 2))#14,add(double(2), 1)#15]
   +- Project [pythonUDF0#16 AS double(add(1, 2))#14,pythonUDF0#18 AS add(double(2), 1)#15]
      +- EvaluatePython [add(pythonUDF1#17, 1)], [pythonUDF0#18]
         +- EvaluatePython [double(add(1, 2)),double(2)], [pythonUDF0#16,pythonUDF1#17]
            +- OneRowRelation$

== Optimized Logical Plan ==
Project [pythonUDF0#16 AS double(add(1, 2))#14,pythonUDF0#18 AS add(double(2), 1)#15]
+- EvaluatePython [add(pythonUDF1#17, 1)], [pythonUDF0#18]
   +- EvaluatePython [double(add(1, 2)),double(2)], [pythonUDF0#16,pythonUDF1#17]
      +- OneRowRelation$

== Physical Plan ==
WholeStageCodegen
:  +- Project [pythonUDF0#16 AS double(add(1, 2))#14,pythonUDF0#18 AS add(double(2), 1)#15]
:     +- INPUT
+- !BatchPythonEvaluation [add(pythonUDF1#17, 1)], [pythonUDF0#16,pythonUDF1#17,pythonUDF0#18]
   +- !BatchPythonEvaluation [double(add(1, 2)),double(2)], [pythonUDF0#16,pythonUDF1#17]
      +- Scan OneRowRelation[]
```
## How was this patch tested?

Added new tests.

Using the following script to benchmark 1, 2 and 3 udfs,

```
df = sqlContext.range(1, 1 << 23, 1, 4)
double = F.udf(lambda x: x * 2, LongType())
print df.select(double(df.id)).count()
print df.select(double(df.id), double(df.id + 1)).count()
print df.select(double(df.id), double(df.id + 1), double(df.id + 2)).count()
```

Here is the results:

| N | Before | After | speed up |
| --- | --- | --- | --- |
| 1 | 22 s | 7 s | 3.1X |
| 2 | 38 s | 13 s | 2.9X |
| 3 | 58 s | 16 s | 3.6X |

This benchmark ran locally with 4 CPUs. For 3 UDFs, it launched 12 Python before before this patch, 4 process after this patch. After this patch, it will use less memory for multiple UDFs than before (less buffering).

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 1]
```diff
@@ -77,22 +77,30 @@ private[spark] case class PythonFunction(
     broadcastVars: JList[Broadcast[PythonBroadcast]],
     accumulator: Accumulator[JList[Array[Byte]]])
 
+
+object PythonRunner {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: This should be `private[spark]`.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 2]
```diff
@@ -77,22 +77,30 @@ private[spark] case class PythonFunction(
...
+  }
+}
+
 /**
- * A helper class to run Python UDFs in Spark.
+ * A helper class to run Python mapPartition/UDFs in Spark.
  */
 private[spark] class PythonRunner(
-    funcs: Seq[PythonFunction],
+    funcs: Seq[Seq[PythonFunction]],
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: This type is a little strange, so do you mind adding a scaladoc comment to explain what the two levels of nesting correspond to?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 3]
```diff
@@ -77,22 +77,30 @@ private[spark] case class PythonFunction(
...
     bufferSize: Int,
     reuse_worker: Boolean,
-    rowBased: Boolean)
+    isUDF: Boolean,
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Similarly, do you mind adding scaldoc for these two new parameters?

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 4]
```diff
@@ -284,11 +292,22 @@ private[spark] class PythonRunner(
...
         // Serialized command:
-        dataOut.writeInt(if (rowBased) 1 else 0)
-        dataOut.writeInt(funcs.length)
-        funcs.foreach { f =>
-          dataOut.writeInt(f.command.length)
-          dataOut.write(f.command)
+        if (isUDF) {
+          dataOut.writeInt(1)
+          dataOut.writeInt(funcs.length)
+          funcs.zip(numArgs).foreach { case (fs, numArg) =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: Since correctness relies on `funcs.length == numArgs.length`, do you mind adding a `require` at the start of the constructor to enforce this?

File: python/pyspark/worker.py [hunk 5]
```diff
@@ -59,7 +59,49 @@ def read_command(serializer, file):
 
 def chain(f, g):
     """chain two function together """
-    return lambda x: g(f(x))
+    return lambda *a: g(f(*a))
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: Woah, didn't know that you could do varargs lambdas. Cool!

File: python/pyspark/worker.py [hunk 6]
```diff
@@ -59,7 +59,49 @@ def read_command(serializer, file):
...
+            row_func = f
+        else:
+            row_func = chain(row_func, f)
+    # the last returnType will be the return type of UDF
+    return num_arg, wrap_udf(row_func, return_type)
+
+
+def read_udfs(pickleSer, infile):
+    num_udfs = read_int(infile)
+    udfs = []
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: It looks like `udfs` holds `(something, something, udf)` triples. Mind adding a line-comment here to say what the first two components of the tuple correspond to?

File: python/pyspark/worker.py [hunk 7]
```diff
@@ -107,21 +149,10 @@ def main(infile, outfile):
...
-            for i in range(num_commands):
-                f, returnType, deserializer = read_command(pickleSer, infile)
-                if row_func is None:
-                    row_func = f
-                else:
-                    row_func = chain(row_func, f)
-            serializer = deserializer
-            func = lambda _, it: map(lambda x: returnType.toInternal(row_func(*x)), it)
+        is_udf = read_int(infile)
+        if is_udf:
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: I'd _maybe_ call this `is_sql_udf` just to make it clearer that this is part of PySpark SQL support, but I don't feel strongly about this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala [hunk 8]
```diff
@@ -69,11 +69,14 @@ case class BatchPythonEvaluation(udf: PythonUDF, output: Seq[Attribute], child:
...
 
-      val (pyFuncs, children) = collectFunctions(udf)
+      val (pyFuncs, children) = udfs.map(collectFunctions).unzip
+      val numArgs = children.map(_.length)
 
       val pickle = new Pickler
-      val currentRow = newMutableProjection(children, child.output)()
-      val fields = children.map(_.dataType)
+      // flatten all the arguments
+      val allChildren = children.flatMap(x => x)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: Quick clarification: if I have a function like `select udf(x), udf2(x), udf3(x), udf4(x) from ...`, we'll send the `x` column's value four times to PySpark? I know that we have a conceptually similar problem when we're evaluating multiple aggregates in parallel in JVM Spark SQL, but in that case I think we only project each column once and end up rebinding the references / offsets to reference the single copy.

My hunch is that this extra copy isn't a huge perf. issue compared to the slow multiple-Python-UDF evaluation strategy we were using before, so I think it's fine to leave this for now. If it does become a problem, we could optimize later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala [hunk 9]
```diff
@@ -89,19 +92,30 @@ case class BatchPythonEvaluation(udf: PythonUDF, output: Seq[Attribute], child:
...
+        StructType(udfs.map(u => StructField("", u.dataType, u.nullable)))
+      }
       val resultProj = UnsafeProjection.create(output, output)
 
       outputIterator.flatMap { pickedResult =>
         val unpickledBatch = unpickle.loads(pickedResult)
         unpickledBatch.asInstanceOf[java.util.ArrayList[Any]].asScala
       }.map { result =>
-        row(0) = EvaluatePython.fromJava(result, udf.dataType)
+        val row = if (udfs.length == 1) {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: Rather than evaluating this `if` condition for every row, could we lift this out of the `map` and perform it once while building the RDD DAG? i.e. assign the result of line 108 to a variable and have the `if` be the last return value of this block?

Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: If you do this, you could reduce the scope of the `mutableRow` created up on line 99, too.

File: python/pyspark/worker.py [hunk 10]
```diff
@@ -59,7 +59,49 @@ def read_command(serializer, file):
...
+    for i in range(num_udfs):
+        num_arg, udf = read_single_udf(pickleSer, infile)
+        udfs.append((offset, offset + num_arg, udf))
+        offset += num_arg
+
+    if num_udfs == 1:
+        udf = udfs[0][2]
+
+        # fast path for single UDF
+        def mapper(args):
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: I bet you could even do `mapper = udf` if you wanted to.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 11]
```diff
@@ -47,10 +49,9 @@ private[spark] object ExtractPythonUDFs extends Rule[LogicalPlan] {
     }
   }
 
-  private def collectEvaluatableUDF(expr: Expression): Seq[PythonUDF] = {
-    expr.collect {
-      case udf: PythonUDF if canEvaluateInPython(udf) => udf
-    }
+  private def collectEvaluatableUDF(expr: Expression): Seq[PythonUDF] = expr match {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: Just curious: why the change from `expr.collect` to this more explicit recursion?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 12]
```diff
@@ -59,45 +60,43 @@ private[spark] object ExtractPythonUDFs extends Rule[LogicalPlan] {
 
     case plan: LogicalPlan if plan.resolved =>
       // Extract any PythonUDFs from the current operator.
-      val udfs = plan.expressions.flatMap(collectEvaluatableUDF)
+      val udfs = plan.expressions.flatMap(collectEvaluatableUDF).filter(_.resolved)
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: Shouldn't everything be resolved by the point that we get to physical planning?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 13]
```diff
@@ -59,45 +60,43 @@ private[spark] object ExtractPythonUDFs extends Rule[LogicalPlan] {
...
-              }
-            }
-
-            assert(evaluation != null, "Unable to evaluate PythonUDF.  Missing input attributes.")
-
-            // Trim away the new UDF value if it was only used for filtering or something.
-            logical.Project(
-              plan.output,
-              plan.transformExpressions {
-                case p: PythonUDF if p.fastEquals(udf) => evaluation.resultAttribute
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: I guess we lost the `fastEquals` thing since we're now relying on hashcodes + hash maps. Presumably this isn't a huge perf. issue.

**Comment by JoshRosen** [comment]
Took a quick pass and have a couple of questions to clarify my understanding of this code since I wasn't fully familiar with our old UDF evaluation strategy (for instance, was unfamiliar with our neat trick for handling composed Python UDFs).

File: python/pyspark/worker.py (hunk 10 above)
Comment (davies) [inline python/pyspark/worker.py]: We can't, input of mapper is a tuple, but udf is not

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala (hunk 9 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: Comparing evaluate Python UDF, I think this does not matter, JIT compiler could predict this branch pretty easy.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala (hunk 11 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: `expr.collect` will collect those PythonUDF even the parent is already collected.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala (hunk 12 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: I think so, since the plan is already resolved. Just keeping them here for safe (it exists before my change)

**Comment by davies** [comment]
@JoshRosen I had addressed you comments also pushed some changes to improve the performance.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala (hunk 9 above)
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: Fair enough.

File: python/pyspark/worker.py (hunk 10 above)
Comment (JoshRosen) [inline python/pyspark/worker.py]: Ah, got it. Makes sense.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 14]
```diff
@@ -78,21 +78,41 @@ private[spark] case class PythonFunction(
...
-    funcs: Seq[PythonFunction],
+    funcs: Seq[ChainedPythonFunctions],
     bufferSize: Int,
     reuse_worker: Boolean,
-    rowBased: Boolean)
+    isUDF: Boolean,
+    argOffsets: Array[Array[Int]])
   extends Logging {
 
+  require(funcs.length == argOffsets.length, "numArgs should have the same length as funcs")
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: numArgs -> argOffsets

File: python/pyspark/worker.py [hunk 15]
```diff
@@ -59,7 +59,54 @@ def read_command(serializer, file):
...
+        udfs = {}
+        call_udf = []
+        for i in range(num_udfs):
+            arg_offsets, udf = read_single_udf(pickleSer, infile)
+            udfs['f%d' % i] = udf
+            args = ["a[%d]" % o for o in arg_offsets]
+            call_udf.append("f%d(%s)" % (i, ", ".join(args)))
+        # Create function like this:
+        #   lambda a: (f0(a0), f1(a1, a2), f2(a3))
+        mapper_str = "lambda a: (%s)" % (", ".join(call_udf))
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: Clever! This is a neat trick.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala [hunk 16]
```diff
@@ -69,19 +70,47 @@ case class BatchPythonEvaluation(udf: PythonUDF, output: Seq[Attribute], child:
...
-      val fields = children.map(_.dataType)
-      val schema = new StructType(fields.map(t => new StructField("", t, true)).toArray)
+      val (pyFuncs, inputs) = udfs.map(collectFunctions).unzip
+
+      // flatten all the arguments
+      val allInputs = new ArrayBuffer[Expression]
+      val dataTypes = new ArrayBuffer[DataType]
+      val argOffsets = inputs.map { input =>
+        input.map { e =>
+          if (allInputs.exists(_.semanticEquals(e))) {
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: In the worst-case this loop is N^2, but N is probably pretty small so it probably doesn't matter compared to other perf. issues impacting Python UDFs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala [hunk 17]
```diff
@@ -69,19 +70,47 @@ case class BatchPythonEvaluation(udf: PythonUDF, output: Seq[Attribute], child:
...
+            allInputs.indexWhere(_.semanticEquals(e))
+          } else {
+            allInputs += e
+            dataTypes += e.dataType
+            allInputs.length - 1
+          }
+        }.toArray
+      }.toArray
+      val projection = newMutableProjection(allInputs, child.output)()
+      val schema = StructType(dataTypes.map(dt => StructField("", dt)))
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: I'm not sure whether the name of this struct field will ever appear / get used anywhere, but if you wanted to provide a nicer name I suppose you could pull in the name or .toString from the expression in `allInput`

File: python/pyspark/worker.py [hunk 18]
```diff
@@ -59,7 +59,54 @@ def read_command(serializer, file):
...
+        mapper = eval(mapper_str, udfs)
+
+    func = lambda _, it: map(mapper, it)
+    ser = BatchedSerializer(PickleSerializer(), 100)
```
Comment (JoshRosen) [inline python/pyspark/worker.py]: What serializer did we use before? 100 seems arbitrary here.

Comment (davies) [inline python/pyspark/worker.py]: Before this patch, we use AutoBatchedSerialzier, which could hold thousands of rows (holding more rows in JVM, may cause OOM). 
The `100` is used on Java side.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala (hunk 17 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: It's not used anywhere, we will convert the row into tuple, then passing the items into UDFs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala (hunk 16 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchPythonEvaluation.scala]: Agreed.

**Comment by JoshRosen** [comment]
LGTM pending Jenkins.

**Comment by davies** [comment]
Merging this into master, thanks!
